### PR TITLE
Switch to the new dask CLI

### PR DESCRIPTION
**Pull request by jacobtomlinson**
In https://github.com/dask/dask/pull/9283 we are adding a new top level `dask` CLI command which can be extended by other modules using entry points. A primary motivation here is to improve discoverability by uniting everything under one tool and allowing folks to run `dask --help` and `dask <subcommand> --help` to learn more about the various tools.

This PR adds a new `click` group called `cuda` and moves the `dask-cuda-worker` command under that group with the name `worker`. This means the `dask-cuda-worker` becomes `dask cuda worker` in the new CLI tool.

I haven't made any changes to the existing `dask-cuda-worker` console script so that will still continue to work, but maybe we should add a deprecation warning to it?

I went with this name rather than `dask cuda-worker` because I think it is more readable and also leaves us open to adding more subcommands in the future without cluttering up the top-level `dask` namespace.

```console
$ dask --help             
Usage: dask [OPTIONS] COMMAND [ARGS]...

  Dask command line interface.

Options:
  --version   Show the version and exit.
  -h, --help  Show this message and exit.

Commands:
  cluster    Manage dask clusters.
  cuda       GPU subcommands.
  docs       Open Dask documentation (https://docs.dask.org/) in a web browser.
  info       Information about your dask installation.
  scheduler  Launch a distributed scheduler.
  ssh        Launch a distributed cluster over SSH.
  worker     Launch a distributed worker attached to an existing SCHEDULER.
```

```console
$ dask cuda --help        
Usage: dask cuda [OPTIONS] COMMAND [ARGS]...

  GPU subcommands.

Options:
  -h, --help  Show this message and exit.

Commands:
  worker  Launch a distributed worker with GPUs attached to an existing SCHEDULER.
```

```console
$ dask cuda worker --help  
Usage: dask cuda worker [OPTIONS] [SCHEDULER] [PRELOAD_ARGV]...

  Launch a distributed worker with GPUs attached to an existing SCHEDULER.

  See https://docs.rapids.ai/api/dask-cuda/stable/quickstart.html#dask-cuda-worker for
  info.

Options:
  --host TEXT                     IP address of serving host; should be visible to the
                                  scheduler and other workers. Can be a string (like
                                  ``"127.0.0.1"``) or ``None`` to fall back on the
                                  address of the interface specified by
                                  ``--interface`` or the default interface.
  --nthreads INTEGER              Number of threads to be used for each Dask worker
                                  process.  [default: 1]
...
```

The CLI PR needs to be merged and released before this can be merged.

Fixes https://github.com/rapidsai/dask-cuda/issues/1038

### CODE REVIEW

**Comment by pentschev** [comment]
The idea seems nice, but this is a very intricate component in many scripts and I imagine this will be annoying to abruptly change. Would we be able to keep `dask-cuda-worker` for some time with a `DeprecationWarning` added to it?

**Comment by wence-** [comment]
> The idea seems nice, but this is a very intricate component in many scripts and I imagine this will be annoying to abruptly change. Would we be able to keep `dask-cuda-worker` for some time with a `DeprecationWarning` added to it?

AIUI, the old entry-point continues to exist and will work. I agree that a deprecation warning is a good idea, but not sure exactly how to hook that up.

**Comment by pentschev** [comment]
> AIUI, the old entry-point continues to exist and will work.

Yes, you're right. That was my bad, I didn't pay a lot of attention to the actual changes.

> I agree that a deprecation warning is a good idea, but not sure exactly how to hook that up.

I think we can simply check if `dask_cuda_worker.py` is running without the `cuda` argument and then raise the warning.

**Comment by jacobtomlinson** [comment]
This is still blocked on the upstream PR

**Comment by jacobtomlinson** [comment]
The new CLI has been merged and released so we should be good to merge this now.

I've bumped the minimum version of `dask` and `distributed` to one with the CLI although I'm not sure this is actually necessary as folks can still use `dask-cuda-worker` and the new `dask cuda` CLI namespace is opt-in.

File: requirements.txt
Comment (pentschev) [inline requirements.txt:2]: Not strictly necessary, but not harmful either, we currently pin `>=2022.9.2` so we rely on the latest Dask/Distributed, only at code freeze we will increase the pin.

**Review by pentschev — COMMENTED** [review]
Overall LGTM, thanks @jacobtomlinson . I now wonder, could we have a `dask cuda scheduler` without increasing code duplication too much now? I think this could be useful to have some default overrides, more specifically `DASK_DISTRIBUTED__COMM__UCX__CREATE_CUDA_CONTEXT=True`.

File: setup.py
Comment (jakirkham) [inline setup.py:61]: What would we like to do with this entry point?

**Comment by jakirkham** [comment]
It would be good to add the entry point to the Conda recipe as well

https://github.com/conda-forge/dask-cuda-feedstock/blob/c870fd90153b47cc0fe8060593bf309e27da3b32/recipe/meta.yaml#L17-L18

File: setup.py
Comment (pentschev) [reply setup.py:61]: I assume you're referring to the `dask-config` one, this is from https://github.com/rapidsai/dask-cuda/pull/1006.

Comment (jacobtomlinson) [reply setup.py:61]: Ideally I'd like that functionality to move from `dask-cuda` to `dask-ctl`, especially given that it isn't particularly GPU specific. cc @quasiben 

Alternatively if it is going to stay here I expect it would become `dask cuda config` and we would need to update the click groups accordingly.

Comment (jakirkham) [reply setup.py:61]: Yep we were discussing renaming in issue ( https://github.com/rapidsai/dask-cuda/issues/1038 ). Feel free to share thoughts there 🙂

If it will take a while to move this functionality to `dask-ctl`, renaming may be prudent as an intermediate step in any event

File: dask_cuda/cli.py
Comment (charlesbluca) [inline dask_cuda/cli.py:24]: Not sure if all `cuda.commands` need to reside in the same file as the initial definition to get picked up correctly? Made a few failed attempts to keep the initial structure of the CLI modules before settling on having them all in one shared module, which is similar to what dask-ctl does for its grouped `cluster`/`discovery` commands.

Assuming this is the only way to organize to the CLI commands, we could probably consolidate things up a bit by reusing shared click arguments/options.

File: dask_cuda/cli.py
Comment (charlesbluca) [inline dask_cuda/cli.py:26]: Was thinking a more detailed description like

```suggestion
    """Subcommands to launch or query distributed workers with GPUs."""
```

Could be nice here but happy to hear feedback

File: conda/recipes/dask-cuda/meta.yaml
Comment (charlesbluca) [inline conda/recipes/dask-cuda/meta.yaml:48]: Considered setting these CLI tests programmatically, but don't see any obvious way to grab the new CLI commands from the pyproject.toml.

cc @jakirkham in case there's something I'm missing

File: dask_cuda/cli.py
Comment (pentschev) [inline dask_cuda/cli.py:438]: I guess this is the only argument and without it nothing will happen. Does it make more sense to remove and make the default action (without any non-mandatory arguments) print the configuration?

File: dask_cuda/cli.py
Comment (pentschev) [inline dask_cuda/cli.py:340]: Not a big deal, but are there any reasons for `SCHEDULER` to be all uppercase? Is it referring to something that is normally written in uppercase? If not, I would suggest rewriting this.

Comment (pentschev) [reply dask_cuda/cli.py:340]:
```suggestion
    """Launch a distributed worker with GPUs attached to an existing dask scheduler.
```

File: dask_cuda/cli.py
Comment (pentschev) [inline dask_cuda/cli.py:31]:
```suggestion
    with the equivalent ``dask scheduler`` option.""",
```

Comment (pentschev) [reply dask_cuda/cli.py:31]: Not absolutely sure, but I imagine the hyphen here is also not necessary anymore (or deprecated)?

**Review by pentschev — COMMENTED** [review]
Thanks @charlesbluca for picking this up, I've left a few suggestions/questions.

File: conda/recipes/dask-cuda/meta.yaml
Comment (jakirkham) [reply conda/recipes/dask-cuda/meta.yaml:48]: There's some code above for the all `-` ones

The ones with ` `s can't be done programmatically AFAIK

**Comment by ajschmidt8** [comment]
I canceled the latest GH Action run because it looks like the python tests were in-progress for over 4 hours, which is dramatically longer than the usual ~30 minutes that they take.

The logs are below, it looks like there were some errors throughout the tests.

- https://github.com/rapidsai/dask-cuda/actions/runs/3912570589/jobs/6687484798


To prevent this from happening again, I will open a PR to put a reasonable timeout on the Python tests.

This also happened on #1081.

File: dask_cuda/cli.py
Comment (charlesbluca) [reply dask_cuda/cli.py:340]: My assumption was that it was to make it clear that it corresponded to the `SCHEDULER` argument that crops up in the `--help` output:

```
Usage: dask cuda worker [OPTIONS] [SCHEDULER] [PRELOAD_ARGV]...

  Launch a distributed worker with GPUs attached to an existing SCHEDULER.
```

It looks like this is also done for `dask worker`, though I can see how this description could seem somewhat strange from the `dask cuda` output.

Is there a way in click to render documentation for arguments? If so, that could be a good place to give more detail on what the expected input for `SCHEDULER` is; if not, I guess a reasonable compromise is to take your suggestion and dump info about the `SCHEDULER` argument in the subsequent lines of the docstring so that it only appears when running `dask cuda worker`.

File: dask_cuda/cli.py
Comment (charlesbluca) [reply dask_cuda/cli.py:31]: Good catch! Yeah I can go through and switch out instances of the old Dask CLI for the new.

File: dask_cuda/cli.py
Comment (charlesbluca) [reply dask_cuda/cli.py:438]: Yeah was thinking that as well, just wasn't sure if it made more sense to do in a separate PR since the diff here is fairly large already.

(cc @quasiben in case there's something non-obvious this command does when `--get-cluster-configuration` isn't passed as an option)

File: dask_cuda/cli.py
Comment (pentschev) [reply dask_cuda/cli.py:340]: I see your point, it feels a little unusual but I'm ok with either one, if your prefer we can leave it as it is. I have no idea about click rendering documentation, not exactly sure how that would even work on the CLI, so my guess is that this probably doesn't exist.

File: dask_cuda/cli.py
Comment (pentschev) [reply dask_cuda/cli.py:31]: I _think_ @jacobtomlinson had done it all (or almost all), so I hope there won't be many more.

File: conda/recipes/dask-cuda/meta.yaml
Comment (jakirkham) [inline conda/recipes/dask-cuda/meta.yaml:50]: Actually we might be able to do this to consolidate them further

```suggestion
    - dask cuda --help
    {% for e in data.get("project", {}).get("scripts", {}).keys() %}
    - {{ e }} --help
    - {{ e|replace("-", " ") }} --help
    {% endfor %}
```

Comment (charlesbluca) [reply conda/recipes/dask-cuda/meta.yaml:50]: Good catch! Yeah for now we can do this, interested in if we intend to deprecate and eventually remove the old entrypoints, which would force us to hardcode this some time in the future (cc @jacobtomlinson if you've had any discussion around this)

File: dask_cuda/tests/test_dask_cuda_worker.py
Comment (charlesbluca) [inline dask_cuda/tests/test_dask_cuda_worker.py:213]: Changed this because IIUC `pkg_resources.working_set` is listing the installed distributions and not necessarily the importable modules; this becomes an issue if the distribution and module names aren't the same (i.e. one would `conda install pillow` and then `import PIL`), which was causing some failures in CI that seem unrelated to the changes here.

Comment (charlesbluca) [reply dask_cuda/tests/test_dask_cuda_worker.py:213]: Decided to break this off into https://github.com/rapidsai/dask-cuda/pull/1085 since it seems like this issue is blocking CI elsewhere (https://github.com/rapidsai/dask-cuda/pull/1081 for example)

File: conda/recipes/dask-cuda/meta.yaml
Comment (jakirkham) [reply conda/recipes/dask-cuda/meta.yaml]: Yeah that's fair. For that reason wasn't sure whether we would want to do this or not

File: dask_cuda/tests/test_dask_cuda_worker.py
Comment (pentschev) [reply dask_cuda/tests/test_dask_cuda_worker.py:213]: Thanks @charlesbluca , the fix should get merged as soon as tests pass.

Comment (jakirkham) [reply dask_cuda/tests/test_dask_cuda_worker.py:213]: Had a question about using [`importlib.metadata`]( https://docs.python.org/3/library/importlib.metadata.html ) from the Python standard library ( https://github.com/rapidsai/dask-cuda/pull/1085#issuecomment-1385921939 )

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

**Review by jakirkham — APPROVED** [review]

**Review by pentschev — APPROVED** [review]
LGTM, thanks @jacobtomlinson and @charlesbluca for working on this!

**Comment by jakirkham** [comment]
Have noted the corresponding changes for the conda-forge recipe in issue ( https://github.com/conda-forge/dask-cuda-feedstock/issues/17 ). Please feel free to note anything there I may have missed

**Pull request merged**
